In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
from scipy.stats.stats import ttest_ind
import warnings

warnings.filterwarnings('ignore')

In [5]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'useducation'


engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

education_df = pd.read_sql_query('select * from useducation',con=engine)


# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

In [8]:
fill_list = ["ENROLL", "TOTAL_REVENUE", "FEDERAL_REVENUE",
             "STATE_REVENUE", "LOCAL_REVENUE", "TOTAL_EXPENDITURE",
             "INSTRUCTION_EXPENDITURE", "SUPPORT_SERVICES_EXPENDITURE",
             "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE", "GRADES_PK_G",
             "GRADES_KG_G", "GRADES_4_G", "GRADES_8_G", "GRADES_12_G", "GRADES_1_8_G",
             "GRADES_9_12_G", "GRADES_ALL_G"]

states = education_df["STATE"].unique()

for state in states:
    education_df.loc[education_df["STATE"] == state, fill_list] = education_df.loc[education_df["STATE"] == state, fill_list].interpolate()

# we drop the null values after interpolation
education_df.dropna(inplace=True)

Don't forget to apply the most suitable missing value filling techniques you applied in the previous checkpoints to the data. You should provide your answers to the following questions after you handled the missing values.

Say, we want to understand the relationship between the expenditures of the governments and the students' overall success in the math and reading.

### 1. Create a new score variable from the weighted averages of all score variables in the datasets. Notice that the number of students in the 4th grade isn't the same as the number of students in the 8th grade. So, you should appropriately weigh the scores.

In [9]:
education_df["overall_score"] = (education_df["GRADES_4_G"]*((education_df["AVG_MATH_4_SCORE"] + education_df["AVG_READING_4_SCORE"])*0.5) + education_df["GRADES_8_G"]
                                 * ((education_df["AVG_MATH_8_SCORE"] + education_df["AVG_READING_8_SCORE"])*0.5))/(education_df["GRADES_4_G"] + education_df["GRADES_8_G"])

The weighted averages are calculated with the number of students in each grade because as noted, there are not the same amount of students in each grade.

### 2. What are the correlations between this newly created score variable and the expenditure types? Which 1 of the expenditure types is more correlated than the others?

In [10]:
education_df[["overall_score", "TOTAL_EXPENDITURE", "INSTRUCTION_EXPENDITURE",
              "SUPPORT_SERVICES_EXPENDITURE", "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE"]].corr()

overall_score  TOTAL_EXPENDITURE  \
overall_score                      1.000000           0.074155   
TOTAL_EXPENDITURE                  0.074155           1.000000   
INSTRUCTION_EXPENDITURE            0.087386           0.990255   
SUPPORT_SERVICES_EXPENDITURE       0.079002           0.992988   
OTHER_EXPENDITURE                  0.010959           0.943702   
CAPITAL_OUTLAY_EXPENDITURE        -0.005389           0.925784   

                              INSTRUCTION_EXPENDITURE  \
overall_score                                0.087386   
TOTAL_EXPENDITURE                            0.990255   
INSTRUCTION_EXPENDITURE                      1.000000   
SUPPORT_SERVICES_EXPENDITURE                 0.974399   
OTHER_EXPENDITURE                            0.902173   
CAPITAL_OUTLAY_EXPENDITURE                   0.884798   

                              SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  \
overall_score                                     0.079002           0.010959   
TOTAL_EXPENDITURE                                 0.992988           0.943702   
INSTRUCTION_EXPENDITURE                           0.974399           0.902173   
SUPPORT_SERVICES_EXPENDITURE                      1.000000           0.952282   
OTHER_EXPENDITURE                                 0.952282           1.000000   
CAPITAL_OUTLAY_EXPENDITURE                        0.911245           0.918188   

                              CAPITAL_OUTLAY_EXPENDITURE  
overall_score                                  -0.005389  
TOTAL_EXPENDITURE                               0.925784  
INSTRUCTION_EXPENDITURE                         0.884798  
SUPPORT_SERVICES_EXPENDITURE                    0.911245  
OTHER_EXPENDITURE                               0.918188  
CAPITAL_OUTLAY_EXPENDITURE                      1.000000

The instruction expenditure is the most correlated of the expenditures to the overall score variable that we created.

### 3. Now, apply PCA to the 4 expenditure types. How much of the total variance is explained by the 1st component?

In [12]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

X = education_df[["INSTRUCTION_EXPENDITURE", "SUPPORT_SERVICES_EXPENDITURE",
                  "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE"]]

X = StandardScaler().fit_transform(X)

sklearn_pca = PCA(n_components=1)
education_df["pca_1"] = sklearn_pca.fit_transform(X)

print(sklearn_pca.explained_variance_ratio_)

[0.9430175]


Roughly 94% of the total variance is explained by the 1st component.

### 4. What is the correlation between the overall score variable and the 1st principal component?

In [13]:
education_df[["overall_score", "pca_1", "TOTAL_EXPENDITURE", "INSTRUCTION_EXPENDITURE",
              "SUPPORT_SERVICES_EXPENDITURE", "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE"]].corr()

overall_score     pca_1  TOTAL_EXPENDITURE  \
overall_score                      1.000000  0.044596           0.074155   
pca_1                              0.044596  1.000000           0.992054   
TOTAL_EXPENDITURE                  0.074155  0.992054           1.000000   
INSTRUCTION_EXPENDITURE            0.087386  0.968606           0.990255   
SUPPORT_SERVICES_EXPENDITURE       0.079002  0.988345           0.992988   
OTHER_EXPENDITURE                  0.010959  0.971343           0.943702   
CAPITAL_OUTLAY_EXPENDITURE        -0.005389  0.955792           0.925784   

                              INSTRUCTION_EXPENDITURE  \
overall_score                                0.087386   
pca_1                                        0.968606   
TOTAL_EXPENDITURE                            0.990255   
INSTRUCTION_EXPENDITURE                      1.000000   
SUPPORT_SERVICES_EXPENDITURE                 0.974399   
OTHER_EXPENDITURE                            0.902173   
CAPITAL_OUTLAY_EXPENDITURE                   0.884798   

                              SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  \
overall_score                                     0.079002           0.010959   
pca_1                                             0.988345           0.971343   
TOTAL_EXPENDITURE                                 0.992988           0.943702   
INSTRUCTION_EXPENDITURE                           0.974399           0.902173   
SUPPORT_SERVICES_EXPENDITURE                      1.000000           0.952282   
OTHER_EXPENDITURE                                 0.952282           1.000000   
CAPITAL_OUTLAY_EXPENDITURE                        0.911245           0.918188   

                              CAPITAL_OUTLAY_EXPENDITURE  
overall_score                                  -0.005389  
pca_1                                           0.955792  
TOTAL_EXPENDITURE                               0.925784  
INSTRUCTION_EXPENDITURE                         0.884798  
SUPPORT_SERVICES_EXPENDITURE                    0.911245  
OTHER_EXPENDITURE                               0.918188  
CAPITAL_OUTLAY_EXPENDITURE                      1.000000

The correlation is about 4.46%, making it a lower correlation than that of the overall score and instruction expenditure, support service expenditure, and total expenditure variables.

### 5. If you were to choose the best variables for your model, would you prefer using the 1st principal component instead of the expenditure variables? Why?

The highest correlation here is the instruction expenditure variable with the overall score, which is about twice the correlation between the overall score and the first principal component. For this reason, using the instruction expenditure variable makes more sense than using the PCA. PCA works best when the correlation between the variables are less than and equal to 0.8. Correlations that are too high can result in components that are unstable and, therefore, change a lot in response to very small changes in the data.